File: ts/src/websocketUtils/awap.ts
import * as $dara from '@darabonba/typescript';

export type AwapMessageType = string;
export type AwapMessageFormat = 'text' | 'binary';

export interface AwapMessage {
  type: AwapMessageType;
  id: string;
  headers?: { [key: string]: string };
  payload?: any;
  format?: AwapMessageFormat;
}

export interface AwapWebSocketHandler extends $dara.WebSocketHandler {
  handleAwapMessage(session: $dara.WebSocketSessionInfo, message: AwapMessage): Promise<void> | void;
}

export class AbstractAwapWebSocketHandler extends $dara.AbstractWebSocketHandler implements AwapWebSocketHandler {
  handleAwapMessage(_session: $dara.WebSocketSessionInfo, _message: AwapMessage): void {
    throw ErrUseRawMessage;
  }
}

export type AwapMessageOption = (message: AwapMessage) => void;

export function withType(msgType: AwapMessageType): AwapMessageOption {
  return (message: AwapMessage) => {
    message.type = msgType;
  };
}

export function withID(id: string): AwapMessageOption {
  return (message: AwapMessage) => {
    message.id = id;
  };
}

export function withHeader(key: string, value: string): AwapMessageOption {
  return (message: AwapMessage) => {
    message.headers = message.headers || {};
    message.headers[key] = value;
  };
}

export function newAwapMessage(payload: any, ...opts: AwapMessageOption[]): AwapMessage {
  const message: AwapMessage = {
    type: 'UpstreamTextEvent',
    id: generateMessageId(),
    payload,
    headers: {},
  };
  opts.forEach((opt) => opt(message));
  return message;
}

export function buildAwapMessageText(message: AwapMessage): string {
  if (!message) {
    throw new Error('message cannot be nil');
  }

  const now = Date.now();
  const headerLines: string[] = [
    `type:${message.type}`,
    `timestamp:${now}`,
  ];

  if (message.id) {
    headerLines.push(`id:${message.id}`);
  }

  if (message.type === 'AckRequiredTextEvent') {
    headerLines.push('ack:required');
  }

  if (message.headers) {
    Object.keys(message.headers).forEach((key) => {
      headerLines.push(`${key}:${message.headers![key]}`);
    });
  }

  let payloadJSON = '{}';
  if (message.payload !== undefined && message.payload !== null) {
    payloadJSON = JSON.stringify(message.payload);
  }

  return `${headerLines.join('\n')}\n\n${payloadJSON}`;
}

export function buildAwapMessageBinary(message: AwapMessage): Buffer {
  if (!message) {
    throw new Error('message cannot be nil');
  }

  const now = Date.now();
  const headerLines: string[] = [
    `type:${message.type}`,
    `timestamp:${now}`,
  ];

  if (message.id) {
    headerLines.push(`id:${message.id}`);
  }

  if (message.headers) {
    Object.keys(message.headers).forEach((key) => {
      headerLines.push(`${key}:${message.headers![key]}`);
    });
  }

  const headerBytes = Buffer.from(`${headerLines.join('\n')}\n\n`, 'utf8');
  let bodyBytes: Buffer;
  if (message.payload !== undefined && message.payload !== null) {
    if (!Buffer.isBuffer(message.payload)) {
      throw new Error(`payload for binary AWAP message must be Buffer, got ${typeof message.payload}`);
    }
    bodyBytes = message.payload;
  } else {
    bodyBytes = Buffer.alloc(0);
  }

  return Buffer.concat([headerBytes, bodyBytes]);
}

export function parseAwapMessage(message: $dara.WebSocketMessage): AwapMessage {
  const data = message.payload;
  let headerEndIndex = -1;
  for (let i = 0; i < data.length - 1; i++) {
    if (data[i] === 0x0a && data[i + 1] === 0x0a) {
      headerEndIndex = i;
      break;
    }
  }

  if (headerEndIndex === -1) {
    throw new Error('failed to parse AWAP message: no \\n\\n separator found');
  }

  const awapMsg: AwapMessage = {
    type: '',
    id: '',
    headers: {},
  };

  const headerStr = data.slice(0, headerEndIndex).toString('utf8');
  const payloadBytes = data.slice(headerEndIndex + 2);

  headerStr.split('\n').forEach((line) => {
    const trimmed = line.trim();
    if (!trimmed) {
      return;
    }
    const colonIndex = trimmed.indexOf(':');
    if (colonIndex > 0) {
      const key = trimmed.slice(0, colonIndex).trim();
      const value = trimmed.slice(colonIndex + 1).trim();
      switch (key) {
        case 'type':
          awapMsg.type = value;
          break;
        case 'id':
          awapMsg.id = value;
          break;
        default:
          awapMsg.headers![key] = value;
          break;
      }
    }
  });

  if (payloadBytes.length > 0) {
    try {
      awapMsg.payload = JSON.parse(payloadBytes.toString('utf8'));
    } catch (_err) {
      awapMsg.payload = payloadBytes;
    }
  }

  awapMsg.format = message.type === $dara.WebSocketMessageType.Binary ? 'binary' : 'text';
  return awapMsg;
}

function generateMessageId(): string {
  const crypto = require('crypto');
  const hexStr = crypto.randomBytes(16).toString('hex');
  if (hexStr.length >= 32) {
    return hexStr.slice(0, 32);
  }
  const timestamp = `${Date.now()}`;
  const combined = timestamp + hexStr;
  if (combined.length >= 32) {
    return combined.slice(0, 32);
  }
  return combined.padEnd(32, '0');
}

export const ErrUseRawMessage = new Error('use HandleRawMessage');

File: python/alibabacloud_tea_openapi/websocket_utils/awap.py
import json
import os
from typing import Any, Callable, Dict, List, Optional

from darabonba.websocket import (
    AbstractWebSocketHandler,
    WebSocketMessage,
    WebSocketMessageType,
    WebSocketSessionInfo,
)

from ._constants import ERR_USE_RAW_MESSAGE


AwapMessage = Dict[str, Any]
AwapMessageType = str
AwapMessageOption = Callable[[AwapMessage], None]


class AwapWebSocketHandler(AbstractWebSocketHandler):
    def handle_awap_message(self, session: WebSocketSessionInfo, message: AwapMessage) -> None:
        raise ERR_USE_RAW_MESSAGE


def with_type(msg_type: str) -> AwapMessageOption:
    def apply(message: AwapMessage) -> None:
        message['type'] = msg_type

    return apply


def with_id(message_id: str) -> AwapMessageOption:
    def apply(message: AwapMessage) -> None:
        message['id'] = message_id

    return apply


def with_header(key: str, value: str) -> AwapMessageOption:
    def apply(message: AwapMessage) -> None:
        message.setdefault('headers', {})[key] = value

    return apply


def new_awap_message(payload: Any, *opts: AwapMessageOption) -> AwapMessage:
    message: AwapMessage = {
        'type': 'UpstreamTextEvent',
        'id': _generate_message_id(),
        'payload': payload,
        'headers': {},
    }
    for opt in opts:
        opt(message)
    return message


def build_awap_message_text(message: AwapMessage) -> str:
    if message is None:
        raise ValueError('message cannot be nil')

    header_lines = [
        f"type:{message['type']}",
        f"timestamp:{int(__import__('time').time() * 1000)}",
    ]
    if message.get('id'):
        header_lines.append(f"id:{message['id']}")
    if message.get('type') == 'AckRequiredTextEvent':
        header_lines.append('ack:required')
    for key, value in (message.get('headers') or {}).items():
        header_lines.append(f'{key}:{value}')

    payload = message.get('payload')
    payload_json = '{}' if payload is None else json.dumps(payload, separators=(',', ':'))
    return '\n'.join(header_lines) + '\n\n' + payload_json


def build_awap_message_binary(message: AwapMessage) -> bytes:
    if message is None:
        raise ValueError('message cannot be nil')

    header_lines = [
        f"type:{message['type']}",
        f"timestamp:{int(__import__('time').time() * 1000)}",
    ]
    if message.get('id'):
        header_lines.append(f"id:{message['id']}")
    for key, value in (message.get('headers') or {}).items():
        header_lines.append(f'{key}:{value}')

    header_bytes = ('\n'.join(header_lines) + '\n\n').encode('utf-8')
    payload = message.get('payload')
    if payload is None:
        body_bytes = b''
    elif isinstance(payload, bytes):
        body_bytes = payload
    else:
        raise ValueError(
            f'payload for binary AWAP message must be bytes, got {type(payload).__name__}'
        )
    return header_bytes + body_bytes


def parse_awap_message(message: WebSocketMessage) -> AwapMessage:
    data = message.payload
    header_end_index = -1
    for i in range(len(data) - 1):
        if data[i] == 0x0A and data[i + 1] == 0x0A:
            header_end_index = i
            break
    if header_end_index == -1:
        raise ValueError('failed to parse AWAP message: no \\n\\n separator found')

    awap_msg: AwapMessage = {
        'type': '',
        'id': '',
        'headers': {},
    }
    header_str = data[:header_end_index].decode('utf-8')
    payload_bytes = data[header_end_index + 2:]

    for line in header_str.split('\n'):
        trimmed = line.strip()
        if not trimmed:
            continue
        colon_index = trimmed.find(':')
        if colon_index > 0:
            key = trimmed[:colon_index].strip()
            value = trimmed[colon_index + 1:].strip()
            if key == 'type':
                awap_msg['type'] = value
            elif key == 'id':
                awap_msg['id'] = value
            else:
                awap_msg['headers'][key] = value

    if payload_bytes:
        try:
            awap_msg['payload'] = json.loads(payload_bytes.decode('utf-8'))
        except (json.JSONDecodeError, UnicodeDecodeError):
            awap_msg['payload'] = payload_bytes

    awap_msg['format'] = 'binary' if message.type == WebSocketMessageType.Binary else 'text'
    return awap_msg


def _generate_message_id() -> str:
    hex_str = os.urandom(16).hex()
    if len(hex_str) >= 32:
        return hex_str[:32]
    timestamp = str(int(__import__('time').time() * 1000))
    combined = timestamp + hex_str
    if len(combined) >= 32:
        return combined[:32]
    return combined.ljust(32, '0')
